README.md
# A Bash Logger

Often, implementing less-than-minimal bash scripts, it would be nice to have different output levels (for information, warnings, errors, etc.) as well as a unified style to print messages to the user.
This need made me implement this logger, which serves for this purpose.
Since I like colorful output, a different color was assigned to each level and your terminal should support 256-colors, which you can simply check trying the following command.

```bash
printf '\n\e[1;38;5;202;48;5;11m Is this bold orange on yellow background? \e[0m\n\n'
```

If no strange symbols are displayed and you see colors, than you are in a good shape to use the logger! :smile:  
If strange symbols appears, have a look to [this however not fully comprehensive list](https://misc.flogisoft.com/bash/tip_colors_and_formatting#terminals_compatibility) to check out which terminals support 256-colors.

## How to use it

If you want simply try it out, you can clone this repository and run the `TestLogger.bash` script as such, or prepending `VERBOSE=TRACE` to activate all levels of output (or something according to the table below).

If you plan to include the logger in your project, feel free to do so.
It should be enough to add the `Logger.bash` file to your codebase.
In any case, please,
* give credit to the author e.g. in the `README` of your repository;
* leave the license and copyright notice in the `Logger.bash` file, although you might prepend there the header of your codebase.

---

### Instructions for the user

This logger has been implemented with the aim to provide functionality once sourced in another script.
Therefore, it simply contains a bunch of functions which become available into your script, once the logger has been sourced.
By default, errors messages (`ERROR`, `FATAL` and `INTERNAL`) are printed to standard error, while the other ones to a customizable file descriptor.

1. Source the logger in your script via `source /path/to/Logger.bash [OPTIONS...]`.
   #### Available options at source time:

   **`--fd N`**  
   Use `N` as logger output file descriptor, with `0<N<255` (default: `N=9`).

   **`--default-exit-code N`**  
   Use `N` as default exit code for `FATAL` and `INTERNAL`, with `0≤N≤255` (default: `N=1`). The `exit_code` variable can be used to change this either globally or on a per-case basis (see below).
1. Use either of the provided functions when you wish to print something to the user:
   * `PrintTrace` or `Print_Trace`
   * `PrintDebug` or `Print_Debug`
   * `PrintInfo` or `Print_Info`
   * `PrintAttention` or `Print_Attention`
   * `PrintWarning` or `Print_Warning`
   * `PrintError` or `Print_Error`
   * `PrintFatalAndExit` or `Print_Fatal_And_Exit`
   * `PrintInternalAndExit` or `Print_Internal_And_Exit`
   
   Note that either of the two syntaxes produces identical results and the reason both are provided is that then you might pick up that with a name closer to the convention used in your codebase.
1. Each function must be followed by at least one argument (possibly empty).
   If the only provided argument is made of `\n` only, that amount of empty lines is printed to the corresponding file descriptor and nothing more (e.g. no label).
   Each argument is printed on a new line, prepending the level label to the first argument and just spaces to the following ones.
   Any argument can be prepended by `--emph` and this makes the following string be emphasized, i.e. printed in a standing out color (run the `TestLogger.bash` script to get an example).
   An emphasised string is considered to be part of the _same_ line and, strictly speaking, it means that not all arguments are printed on new lines.
   This is surely a desired and intuitive default behaviour and manual `\n` can be used to cope with different wishes.
   Note that a trailing `--emph` argument will be printed literally in the default level color, as well as if `--emph` is given twice in a row (this time is printed in the emphasized color).

   Each argument is printed to the chosen fd (which duplicates the standard output when the logger is sourced).
   **This means that you can use the logger in a function and still capture its standard output via the `$(call_to_function)` syntax.**
   This works because command expansion `$()` captures the standard output only!

   The logger functions should correctly deal with both literal `\n` and interpreted `$'\n'`.
   These will split the message and indent the second line appropriately.
   However, this should rarely be used as multiple arguments achieve the same and avoid the user having to manually deal with endlines.
   Resulting code is therefore also more readable and a long bash line can be split more easily if it is made of multiple arguments.
   A use case for a literal `\n` might be at the end of an argument when the following one is emphasized and the user wishes the latter to start on a new line.
1. By default, both `PrintFatalAndExit` and `PrintInternalAndExit` will exit with exit code 1 (general failure).
   You can however use the `exit_code` variable to change this either globally or on a per-case basis.
   E.g. `exit_code=42 PrintFatalAndExit "Oh no!"` will use 42 as exit code (and the `exit_code` variable will not be changed in the calling environment).
1. You can pass either of the following command-line argument to each function to slightly change its behaviour.
   In this case separate the message from the options via `--`.
   * `-l` &ensp; :arrow_right: &ensp; to suppress label printing (space is printed instead)
   * `-n` &ensp; :arrow_right: &ensp; to suppress last endline (useful e.g. to ask interactive question to user)
   * `-d` &ensp; :arrow_right: &ensp; to avoid restoring default color and font at the end.
   * `--` &ensp; :arrow_right: &ensp; to separate options from message
   
   For example, `PrintInfo -l -- "Hello world!"` will print the message in green with some space in front, but without label.
1. The output levels have an internal hierarchy and it is possible to choose what to print.
   Use the `VERBOSE` environment variable when running your script to activate or suppress the output level printing.

   | VERBOSE | Levels that are printed |
   | :-----: | :---------------------- |
   | `INTERNAL`         | Internal, Fatal |
   | `0` or `FATAL`     | Internal, Fatal |
   | `1` or `ERROR`     | Internal, Fatal, Error |
   | `2` or `WARNING`   | Internal, Fatal, Error, Warning |
   | `3` or `ATTENTION` | Internal, Fatal, Error, Warning, Attention |
   | `4` or `INFO`      | Internal, Fatal, Error, Warning, Attention, Info |
   | `5` or `DEBUG`     | Internal, Fatal, Error, Warning, Attention, Info, Debug |
   | `6` or `TRACE`     | Internal, Fatal, Error, Warning, Attention, Info, Debug, Trace |
   | Anything else      | Internal, Fatal, Error, Warning, Attention, Info |
   
   For example, running your script via `VERBOSE=DEBUG ./yourScript` will activate also the debug level.

## Features planned to be added

 - [x] Implement mechanism to highlight part of message
 - [ ] Add command-line option to customize at source time:
    - [x] Offer `fd` customization
    - [x] Offer default exit code customization
    - [ ] Offer label length customization (now fixed)
    - [ ] Offer level color customization (now fixed)
    - [ ] Offer date activation (now only the level label is printed)
 - [ ] Allow literal `--` to be used and be printed (now it separates options from strings)
 - [ ] Explore the possibility to implement a "level factory" to give possibility to the user to create their levels
 
## A similar logger in C++

The idea to implement this logger in this way came to my mind from the fact that I already sometimes use a similar tool in `C++` codebases, which is called **Einhard** and which is [freely available](https://gitlab.com/Marix/Einhard).
Check it out!


CHANGELOG.md
# Changelog

All notable changes to the `BashLogger` will be documented in this file.

Strictly speaking, this project does not adhere to [Semantic Versioning](http://semver.org/spec/v2.0.0.html), but it follows some aspects inspired from it.
In particular, given a version number `X.Y`,
 - `Y` is incremented for minor changes (e.g. bug fixes) and minor new features and
 - `X` for major ones.

In this project, version `1.0` is considered to be reached when the logger starts to be _complete_ in view of the developers.
Any of the releases before is stable and usable, though.

### Legend

 * :new: New feature
 * :white_check_mark: Enhancement
 * :sos: Bug fix
 * :x: Removed feature

---

## [Unreleased]

 * :white_check_mark: Let single argument made of endlines only be accepted by logger functions
 * :sos: Make all lines be correctly indented when `\n` is used in function arguments
 * :white_check_mark: Use 9 as default output file descriptor (as Bash discourages larger ones)

---

## [Version 0.2] &nbsp;&nbsp; <sub><sup>23 June 2023</sub></sup>

* :white_check_mark: Provide additional user interface with `_`-separated words in function names (instead of camel-case)
* :new: Implement mechanism to emphasize part of the message
* :new: Allow user to choose the default exit code to be used from the logger for fatal errors
* :new: Allow user to choose at source time the output file descriptor to be used
* :sos: Fix logger to be used when bash `-u` option is activated

---

## [Version 0.1] &nbsp;&nbsp; <sub><sup>19 December 2019</sub></sup>

This has been the first release of the repository.


[Unreleased]: https://github.com/AxelKrypton/BashLogger/compare/v0.2...HEAD
[Version 0.2]: https://github.com/AxelKrypton/BashLogger/releases/tag/v0.2
[Version 0.1]: https://github.com/AxelKrypton/BashLogger/releases/tag/v0.1


TestLogger.bash
#!/usr/bin/env bash
#
#  Copyright (c) 2019 Alessandro Sciarra <sciarra@itp.uni-frankfurt.de>
#
#  This file is part of BashLogger.
#
#  BashLogger is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#
#  BashLogger is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with BashLogger. If not, see <https://www.gnu.org/licenses/>.
#

trap 'printf "\n"' EXIT
source Logger.bash "$@" || exit 1

Print_Trace                  '\nTrace message '         --emph 'highlighted' ' VS normal'
Print_Debug                  '\nDebug message '         --emph 'highlighted' ' VS normal'
Print_Info                   '\nInformation message '   --emph 'highlighted' ' VS normal'
Print_Attention              '\n' 'Attention message '  --emph 'highlighted' ' VS normal'
Print_Warning                '\n' 'Warning message '    --emph 'highlighted' ' VS normal'
Print_Error                  '\n' 'Error '              --emph 'highlighted' ' VS normal'
(
    Print_Fatal_And_Exit     '\n' 'Fatal error exit! '  --emph 'highlighted' ' VS normal'
    Print_Internal_And_Exit  '\n' 'Developer error '    --emph 'highlighted' ' VS normal'
)

if [[ "${TEST}" = '' ]]; then
    exit 0
fi

Print_Info               '\nTry 100% ' --emph 'highlighted' ' VS normal'
Print_Info               $'\nText with literal\nendline' --emph $'\nalso in\nemphasized string'
Print_Info               'Text with trailing \\n and following --emph text\n' --emph '\nwith \\n\nalso in it'
Print_Info               'Trailing' '--emph'
Print_Info               --emph '--emph'
Print_Info               --emph
Print_Info               ''
Print_Info               '\n\n'


Logger.bash
#
#  Copyright (c) 2019,2023
#    Alessandro Sciarra <sciarra@itp.uni-frankfurt.de>
#
#  This file is part of BashLogger.
#
#  BashLogger is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#
#  BashLogger is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with BashLogger. If not, see <https://www.gnu.org/licenses/>.
#
#----------------------------------------------------------------------------------------
#
# The logger will print output to the chosen file descriptor (by default 9). This is
# done (instead of simply let it print to standard output) to be able to use the logger
# in functions that "return" by printing to stdout and that are meant to be called in $().
#
# ATTENTION: It might be checked if the chosen fd exists and in case open it in the Logger
#            function itself. However, if the first Logger call is done in a subshell, then
#            the chosen fd would not be open globally for the script and following calls to
#            the Logger would fail. Hence, we open the fd at source time and not close it
#            in the Logger.
#
# NOTE: Nothing is done if this file is executed and not sourced!
#
if [[ "${BASH_SOURCE[0]}" != "${0}" ]]; then
    BSHLGGR_outputFd=9
    BSHLGGR_defaultExitCode=1
    while [[ $# -gt 0 ]]; do
        case "$1" in
            --fd)
                if [[ ! $2 =~ ^[1-9][0-9]*$ ]] || (($2 > 254)); then
                    printf "Error sourcing BashLogger. '$1' option needs an integer value between 1 and 254.\n"
                    return 1
                else
                    BSHLGGR_outputFd=$2
                    shift 2
                fi
                ;;
            --default-exit-code)
                if [[ ! $2 =~ ^[0-9]+$ ]] || (($2 > 255)); then
                    printf "Error sourcing BashLogger. '$1' option needs an integer value between 0 and 255.\n"
                    return 1
                else
                    BSHLGGR_defaultExitCode=$2
                    shift 2
                fi
                ;;
            *)
                printf "Error sourcing BashLogger. Unknown option '$1'.\n"
                return 1
                ;;
        esac
    done
    # Probably redundant check, but we want guarantee that 'eval' is safe to use here
    # e.g. that the BSHLGGR_outputFd cannot be set to '; rm -rf /; #' (AAARGH!).
    if [[ ! ${BSHLGGR_outputFd} =~ ^[1-9][0-9]*$ ]]; then
        printf "Unexpected error sourcing BashLogger. Please contact developers.\n"
        return 1
    else
        eval "exec ${BSHLGGR_outputFd}>&1"
    fi
    readonly BSHLGGR_outputFd BSHLGGR_defaultExitCode
fi

function PrintTrace()
{
    __static__Logger 'TRACE' "$@"
}
function Print_Trace()
{
    PrintTrace "$@"
}

function PrintDebug()
{
    __static__Logger 'DEBUG' "$@"
}
function Print_Debug()
{
    PrintDebug "$@"
}

function PrintInfo()
{
    __static__Logger 'INFO' "$@"
}
function Print_Info()
{
    PrintInfo "$@"
}

function PrintAttention()
{
    __static__Logger 'ATTENTION' "$@"
}
function Print_Attention()
{
    PrintAttention "$@"
}

function PrintWarning()
{
    __static__Logger 'WARNING' "$@"
}
function Print_Warning()
{
    PrintWarning "$@"
}

function PrintError()
{
    __static__Logger 'ERROR' "$@"
}
function Print_Error()
{
    PrintError "$@"
}

function PrintFatalAndExit()
{
    __static__Logger 'FATAL' "$@"
}
function Print_Fatal_And_Exit()
{
    PrintFatalAndExit "$@"
}

function PrintInternalAndExit()
{
    __static__Logger 'INTERNAL' "$@"
}
function Print_Internal_And_Exit()
{
    PrintInternalAndExit "$@"
}

function __static__Logger()
{
    if [[ $# -lt 1 ]]; then
        __static__Logger 'INTERNAL' "${FUNCNAME} called without label!"
    fi
    local label labelLength labelToBePrinted color emphColor finalEndline restoreDefault
    finalEndline='\n'
    restoreDefault='\e[0m'
    labelLength=10
    label="$1"
    shift
    labelToBePrinted=$(printf "%${labelLength}s" "${label}:")
    if [[ ! ${label} =~ ^(INTERNAL|FATAL|ERROR|WARNING|ATTENTION|INFO|DEBUG|TRACE)$ ]]; then
        __static__Logger 'INTERNAL' "${FUNCNAME} called with unknown label '${label}'!"
    fi
    __static__IsLevelOn "${label}" || return 0
    exec 4>&1 # duplicate fd 1 to restore it later
    case "${label}" in
        ERROR | FATAL)
            # ;;& means go on in case matching following patterns
            color='\e[91m'
            ;;&
        INTERNAL)
            color='\e[38;5;202m'
            ;;&
        ERROR | FATAL | INTERNAL)
            emphColor='\e[93m'
            exec 1>&2 # here stdout to stderr!
            ;;
        INFO)
            color='\e[92m'
            emphColor='\e[96m'
            ;;&
        ATTENTION)
            color='\e[38;5;200m'
            emphColor='\e[38;5;141m'
            ;;&
        WARNING)
            color='\e[93m'
            emphColor='\e[38;5;202m'
            ;;&
        DEBUG)
            color='\e[38;5;38m'
            emphColor='\e[38;5;48m'
            ;;&
        TRACE)
            color='\e[38;5;247m'
            emphColor='\e[38;5;256m'
            ;;&
        *)
            exec 1>&"${BSHLGGR_outputFd}" # here stdout to chosen fd
            ;;
    esac
    if __static__IsElementInArray '--' "$@"; then
        while [[ "$1" != '--' ]]; do
            case "$1" in
                -n)
                    finalEndline=''
                    shift
                    ;;
                -l)
                    labelToBePrinted="$(printf "%${labelLength}s" '')"
                    shift
                    ;;
                -d)
                    restoreDefault=''
                    shift
                    ;;
                *)
                    __static__Logger 'INTERNAL' "${FUNCNAME} called with unknown option \"$1\"!"
                    ;;
            esac
        done
        shift
    fi
    # Print out initial new-lines before label suppressing first argument if it was endlines only
    # Note that endlines could be either '\n' or $'\n' and we want to deal with both.
    local new_first_argument
    while [[ $1 =~ ^\\n || "${1:0:1}" = $'\n' ]]; do
        printf '\n'
        if [[ $1 =~ ^\\n ]]; then
            new_first_argument="${1/#\\n/}"
        elif [[ "${1:0:1}" = $'\n' ]]; then
            new_first_argument="${1/#$'\n'/}"
        fi
        if [[ "${new_first_argument}" = '' ]]; then
            if [[ $# -eq 1 ]]; then
                return
            else
                shift
            fi
        else
            set -- "${new_first_argument}" "${@:2}"
        fi
    done
    # Ensure something to print was given
    if [[ $# -eq 0 ]]; then
        __static__Logger 'INTERNAL' "${FUNCNAME} called without message (or with new lines only)!"
    fi
    # Parse all arguments and save messages for later, possibly modified
    local messagesToBePrinted emphNextString lastStringWasEmph indentation message
    local -r const_indentation="${labelToBePrinted//?/ } "
    messagesToBePrinted=()
    emphNextString='FALSE'
    lastStringWasEmph='FALSE'
    indentation=''
    while [[ $# -gt 0 ]]; do
        case "$1" in
            --emph)
                # Here two cases should be handled: either '--emph' is an option or a string
                #   -> it is literal if the last command line option or if following the option
                #   -> it is an option otherwise
                # Use a fallthrough to continue matching in the case construct and use *) case
                # to print. However, shift and continue must be given if '--emph' was an option.
                if [[ $# -eq 1 || ${emphNextString} = 'TRUE' ]]; then
                    :
                else
                    emphNextString='TRUE'
                    shift
                    continue
                fi
                ;;&
            *)
                if [[ ${#messagesToBePrinted[@]} -gt 0 ]]; then
                    indentation="${labelToBePrinted//?/ } "
                fi
                # Set color and replace % by %% for later printf
                if [[ ${emphNextString} = 'TRUE' ]]; then
                    message=$(__static__ReplaceEndlinesInMessage "$1")
                    messagesToBePrinted+=("${emphColor}${message//%/%%}")
                    lastStringWasEmph='TRUE'
                else
                    if [[ ${lastStringWasEmph} = 'FALSE' ]]; then
                        if [[ ${#messagesToBePrinted[@]} -gt 0 ]]; then
                            messagesToBePrinted[-1]+='\n'
                        fi
                    else
                        indentation=''
                    fi
                    message=$(__static__ReplaceEndlinesInMessage "$1")
                    messagesToBePrinted+=("${indentation}${color}${message//%/%%}")
                    lastStringWasEmph='FALSE'
                fi
                emphNextString='FALSE'
                ;;
        esac
        shift
    done
    # Last message has no endline, add 'finalEndline' to it
    messagesToBePrinted[-1]+="${finalEndline}"
    set -- "${messagesToBePrinted[@]}"
    # Print first line
    printf "\e[1m${color}${labelToBePrinted}\e[0m $1"
    shift
    # Print possible additional lines
    while [[ $# -gt 0 ]]; do
        printf "$1"
        shift
    done
    if [[ ${label} = 'INTERNAL' ]]; then
        printf "${labelToBePrinted//?/ } Please, contact developers.\n"
    fi
    printf "${restoreDefault}"
    exec 1>&4- # restore fd 1 and close fd 4 and not close chosen fd (it must stay open, see top of the file!)
    if [[ ${label} =~ ^(FATAL|INTERNAL)$ ]]; then
        exit "${exit_code:-${BSHLGGR_defaultExitCode}}"
    fi
}

function __static__ReplaceEndlinesInMessage()
{
    # Go through the message to be added character by character and prepend indentation
    # to literal '\n' -> Do this with pure bash not to rely on external tools. Using the
    # BASH_REMATCH array would make it tougher to iterate "left to right" hitting the replaced
    # patterns only once in a general fashion.
    #
    # NOTE: It is assumed here that the caller has defined the 'const_indentation' variable.
    local index oneChar twoChars
    message=''
    for ((index = 0; index < ${#1}; index++)); do
        oneChar="${1:${index}:1}"  # one chatacter  from index
        twoChars="${1:${index}:2}" # two chatacters from index
        if [[ "${twoChars}" = '\n' ]]; then
            if [[ ${index} -eq 0 || "${1:$((index - 1)):1}" != '\' ]]; then
                message+="\n${const_indentation}"
                ((index++))
                continue
            fi
        elif [[ "${oneChar}" = $'\n' ]]; then
            message+="\n${const_indentation}"
            continue
        fi
        message+="${oneChar}"
    done
    printf '%s' "${message}"
}

function __static__IsLevelOn()
{
    local label
    label="$1"
    # FATAL and INTERNAL always on
    if [[ ${label} =~ ^(FATAL|INTERNAL)$ ]]; then
        return 0
    fi
    # VERBOSE environment variable defines how verbose the output should be:
    #  - unset, empty, invalid value -> till INFO (no DEBUG TRACE)
    #  - numeric -> till that level (1=ERROR, 2=WARNING, ...)
    #  - string  -> till that level
    local loggerLevels loggerLevelsOn level index
    loggerLevels=([1]='ERROR' [2]='WARNING' [3]='ATTENTION' [4]='INFO' [5]='DEBUG' [6]='TRACE')
    loggerLevelsOn=()
    if [[ ${VERBOSE-} =~ ^[0-9]+$ ]]; then
        loggerLevelsOn=("${loggerLevels[@]:1:VERBOSE}")
    elif [[ ${VERBOSE-} =~ ^(ERROR|WARNING|ATTENTION|INFO|DEBUG|TRACE)$ ]]; then
        for level in "${loggerLevels[@]}"; do
            loggerLevelsOn+=("${level}")
            if [[ ${VERBOSE-} = "${level}" ]]; then
                break
            fi
        done
    elif [[ ${VERBOSE-} =~ ^(FATAL|INTERNAL)$ ]]; then
        loggerLevelsOn=('FATAL')
    else
        loggerLevelsOn=('FATAL' 'ERROR' 'WARNING' 'ATTENTION' 'INFO')
    fi
    for level in "${loggerLevelsOn[@]}"; do
        if [[ ${label} = "${level}" ]]; then
            return 0
        fi
    done
    return 1
}

function __static__IsElementInArray()
{
    # ATTENTION: Since this function is used in the middle of the logger, the
    #            logger cannot be used in this function otherwise fd 4 is closed!
    local elementToBeFound arrayEntry
    elementToBeFound="$1"
    for arrayEntry in "${@:2}"; do
        if [[ "${arrayEntry}" = "${elementToBeFound}" ]]; then
            return 0
        fi
    done
    return 1
}

#-----------------------------------------------------------------#
#Set functions readonly to avoid possible redefinitions elsewhere
readonly -f \
    PrintTrace \
    PrintDebug \
    PrintInfo \
    PrintAttention \
    PrintWarning \
    PrintError \
    PrintFatalAndExit \
    PrintInternalAndExit \
    Print_Trace \
    Print_Debug \
    Print_Info \
    Print_Attention \
    Print_Warning \
    Print_Error \
    Print_Fatal_And_Exit \
    Print_Internal_And_Exit \
    __static__Logger \
    __static__IsLevelOn \
    __static__IsElementInArray
